File: server-shared/src/main/java/rikka/shizuku/server/ClientRecord.java
package rikka.shizuku.server;

import androidx.annotation.Nullable;

import eu.darken.porter.core.CallerIdentity;
import eu.darken.porter.core.ClientCallback;
import moe.shizuku.server.IShizukuApplication;
import rikka.shizuku.server.legacy.LegacyClientCallback;
import rikka.shizuku.server.util.Logger;

public class ClientRecord {

    protected static final Logger LOGGER = new Logger("ClientRecord");

    public final int uid;
    public final int pid;
    /** Null unless the client attached through the Shizuku endpoint. */
    @Nullable
    public final IShizukuApplication client;
    public final ClientCallback callback;
    public final String packageName;
    public final int apiVersion;
    public boolean allowed;

    public ClientRecord(int uid, int pid, IShizukuApplication client, String packageName, int apiVersion) {
        this(new CallerIdentity(uid, pid), new LegacyClientCallback(client), packageName, apiVersion);
    }

    public ClientRecord(CallerIdentity identity, ClientCallback callback, String packageName, int apiVersion) {
        this.uid = identity.uid;
        this.pid = identity.pid;
        this.callback = callback;
        this.client = callback instanceof LegacyClientCallback
                ? ((LegacyClientCallback) callback).application
                : null;
        this.packageName = packageName;
        this.allowed = false;
        this.apiVersion = apiVersion;
    }

    public void dispatchRequestPermissionResult(int requestCode, boolean allowed) {
        try {
            callback.onPermissionResult(requestCode, allowed);
        } catch (Throwable e) {
            LOGGER.w(e, "dispatchRequestPermissionResult failed for client (uid=%d, pid=%d, package=%s)", uid, pid, packageName);
        }
    }
}

File: server-shared/src/main/java/rikka/shizuku/server/ClientRecord.kt
package rikka.shizuku.server

import eu.darken.porter.core.CallerIdentity
import eu.darken.porter.core.ClientCallback
import moe.shizuku.server.IShizukuApplication
import rikka.shizuku.server.legacy.LegacyClientCallback
import rikka.shizuku.server.util.Logger

class ClientRecord(
    identity: CallerIdentity,
    val callback: ClientCallback,
    val packageName: String,
    /**
     * What the client declared when it attached, on the scale of the wire it attached through:
     * the Shizuku API level (-1 before v13) or the Porter protocol version.
     */
    val apiVersion: Int,
) {

    val uid: Int = identity.uid
    val pid: Int = identity.pid

    /** Null unless the client attached through the Shizuku endpoint. */
    val client: IShizukuApplication? = (callback as? LegacyClientCallback)?.application
    var allowed: Boolean = false

    constructor(uid: Int, pid: Int, client: IShizukuApplication, packageName: String, apiVersion: Int) :
        this(CallerIdentity(uid, pid), LegacyClientCallback(client), packageName, apiVersion)

    fun dispatchRequestPermissionResult(requestCode: Int, allowed: Boolean) {
        try {
            callback.onPermissionResult(requestCode, allowed)
        } catch (e: Throwable) {
            LOGGER.w(e, "dispatchRequestPermissionResult failed for client (uid=%d, pid=%d, package=%s)", uid, pid, packageName)
        }
    }

    companion object {
        private val LOGGER = Logger("ClientRecord")
    }
}
